From: Dan Williams <dan.j.williams@intel.com>
Date: Sat, 27 Sep 2008 02:01:20 +0000 (-0700)
Subject: [RFC PATCH] touch_mnt_namespace when the mount flags change
X-git-tag: v2.6.28-rc1~24^2~6
X-git-url: http://git.kernel.org/?p=linux%2Fkernel%2Fgit%2Ftorvalds%2Flinux-2.6.git;a=commitdiff_plain;h=0e55a7cca4b66f625d67b292f80b6a976e77c51b
Patch-mainline: 2.6.28-rc1
References: FATE#304218

[RFC PATCH] touch_mnt_namespace when the mount flags change

Daemons that need to be launched while the rootfs is read-only can now
poll /proc/mounts to be notified when their O_RDWR requests may no
longer end in EROFS.

[[SLES:  This is needed for mdadm-3.0 to be able to handle DDF and IMSM
  arrays early in the boot process ]]

Cc: Kay Sievers <kay.sievers@vrfy.org>
Cc: Neil Brown <neilb@suse.de>
Signed-off-by: Dan Williams <dan.j.williams@intel.com>
Acked-by: Neil Brown <neilb@suse.de>

---

---
 fs/namespace.c |    7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

--- linux-2.6.27-neilb.orig/fs/namespace.c
+++ linux-2.6.27-neilb/fs/namespace.c
@@ -1553,8 +1553,13 @@ static noinline int do_remount(struct na
 	if (!err)
 		nd->path.mnt->mnt_flags = mnt_flags;
 	up_write(&sb->s_umount);
-	if (!err)
+	if (!err) {
 		security_sb_post_remount(nd->path.mnt, flags, data);
+
+		spin_lock(&vfsmount_lock);
+		touch_mnt_namespace(nd->path.mnt->mnt_ns);
+		spin_unlock(&vfsmount_lock);
+	}
 	return err;
 }
 
